# pygenomeworks

Python libraries and utilities for manipulating genomics data

## Installation

```
pip install -r requirements.txt
python setup.py install
```

## Generating a simulated genome

A genome can be simulated without any parameters, to generate a 10Mbp reference with 20x coverage and median read length of 10kbp:

```
genome_simulator --reference_length 2700000 --num_reads 5400 --median_read_length=10000
```

this will generate a 2.7Mbp reference genome with 20x coverage (default errors) reads in two files (run time approx 3 min):

1. `ref.fasta` - the reference genome
2. `reads.fasta` - the corresponding reads

## Reporting assembly quality

`assembly_evaluation` reports on the assembly quality for GFA-format assemblies (e.g those generated by miniasm). It is a wrapper for the [Quast](https://github.com/ablab/quast) tool. This section assumes that we have reads and reference fasta files (e.g generated by `genome_simulator` as demonstrated in the above subsection).

### Step 1. Generate overlaps using minimap2

```
minimap2 -x ava-ont ./reads.fasta ./reads.fasta -t 12> out.paf
```

### Step 2. Assemble the overlaps
```
miniasm -f ./reads.fasta ./out.paf > reads.gfa
```

### Step 3. Evaluate assembly
```
assembly_evaluator --gfa_filepath ./reads.gfa --reference_filepath ./ref.fasta
```

This should produce an output that looks like this:


```
Assembly                     a64b786b_058d_4ba4_a5c8_564c317e8f86
# contigs (>= 0 bp)          1
# contigs (>= 1000 bp)       1
# contigs (>= 5000 bp)       1
# contigs (>= 10000 bp)      1
# contigs (>= 25000 bp)      1
# contigs (>= 50000 bp)      1
Total length (>= 0 bp)       2630124
Total length (>= 1000 bp)    2630124
Total length (>= 5000 bp)    2630124
Total length (>= 10000 bp)   2630124
Total length (>= 25000 bp)   2630124
Total length (>= 50000 bp)   2630124
# contigs                    1
Largest contig               2630124
Total length                 2630124
Reference length             2700000
GC (%)                       55.75
Reference GC (%)             55.86
N50                          2630124
NG50                         2630124
N75                          2630124
NG75                         2630124
L50                          1
LG50                         1
L75                          1
LG75                         1
# misassemblies              0
# misassembled contigs       0
Misassembled contigs length  0
# local misassemblies        0
# scaffold gap ext. mis.     0
# scaffold gap loc. mis.     0
# unaligned mis. contigs     0
# unaligned contigs          0 + 0 part
Unaligned length             0
Genome fraction (%)          100.000
Duplication ratio            0.974
# N's per 100 kbp            0.00
# mismatches per 100 kbp     3.44
# indels per 100 kbp         2064.15
Largest alignment            2630124
Total aligned length         2630124
NA50                         2630124
NGA50                        2630124
NA75                         2630124
NGA75                        2630124
LA50                         1
LGA50                        1
LA75                         1
LGA75                        1

```

indicating that mismatch rate is low, but indel rate is around 2%

### Step 4 (optional). Polish with racon and re-evaluate

1. Convert the GFA to a FA file:

```
awk '/^S/{print ">"$2"\n"$3}' reads.gfa | fold > assembly.fa
```

2. map the reads to the assembly

```
minimap2 assembly.fa reads.fasta > overlaps.paf
```

3. Polish the assembly with Racon

```
racon -c4 -m 8 -x -6 -g -8 -w 500 -t 12 -q -1 reads.fasta overlaps.paf assembly.fa > polished_assembly.fa
```

4. Analyse with Quast:

```
quast.py polished_assembly.fa -r ./ref.fasta
```

5. Looking at the report file (`less ./quast_results/latest/report.txt`) should show results similar to this:


```Assembly                     polished_assembly
# contigs (>= 0 bp)          1
# contigs (>= 1000 bp)       1
# contigs (>= 5000 bp)       1
# contigs (>= 10000 bp)      1
# contigs (>= 25000 bp)      1
# contigs (>= 50000 bp)      1
Total length (>= 0 bp)       2665507
Total length (>= 1000 bp)    2665507
Total length (>= 5000 bp)    2665507
Total length (>= 10000 bp)   2665507
Total length (>= 25000 bp)   2665507
Total length (>= 50000 bp)   2665507
# contigs                    1
Largest contig               2665507
Total length                 2665507
Reference length             2700000
GC (%)                       55.75
Reference GC (%)             55.86
N50                          2665507
NG50                         2665507
N75                          2665507
NG75                         2665507
L50                          1
LG50                         1
L75                          1
LG75                         1
# misassemblies              0
# misassembled contigs       0
Misassembled contigs length  0
# local misassemblies        0
# scaffold gap ext. mis.     0
# scaffold gap loc. mis.     0
# unaligned mis. contigs     0
# unaligned contigs          0 + 0 part
Unaligned length             0
Genome fraction (%)          99.999
Duplication ratio            0.987
# N's per 100 kbp            0.00
# mismatches per 100 kbp     0.59
# indels per 100 kbp         1062.67
Largest alignment            2665501
Total aligned length         2665501
NA50                         2665501
NGA50                        2665501
NA75                         2665501
NGA75                        2665501
LA50                         1
LGA50                        1
LA75                         1
LGA75                        1
```

Mismatch rate has dropped from 2% to 1%
